src/protocol/frame/mod.rs

```diff
@@ -192,7 +196,14 @@ impl FrameCodec {
             }
 
             // Not enough data in buffer.
-            self.in_buffer.reserve(self.header.as_ref().map(|(_, l)| *l as usize).unwrap_or(6));
+            if let Some((_, len)) = &self.header {
+                if !reserved_full_msg_len {
+                    self.in_buffer.reserve(*len as usize);
+                    reserved_full_msg_len = true;
+                }
+            } else {
+                self.in_buffer.reserve(FrameHeader::MAX_SIZE);
+            }
```

Comment:
```suggestion
            let reserve_len = (&self.header).map(|(_, len)| *len as usize).unwrap_or(FrameHeader::MAX_SIZE);
            self.in_buffer.reserve(reserve_len);
```

What do you think of this version? It helps to eliminate `let reserved_full_msg_len`; also, `reserve()` would not re-reserve memory if already reserved, if I'm not mistaken.

Reply:
That's what we had before-ish. It makes sense for messages we'll read in a single chunk.

For large messages though, say 1MB, it will reserve 1MB after each `read_buffer_size`, say 128KiB, chunk read. So we'd end up over-reserving by up to 1MB.

We _could_ only reserve what we need for this chunk, which would solve that. But it seemed more efficient to reserve, and therefore allocate, once instead of multiple times per-chunk of a large message.

That's why I added the flag. For all messages we need only reserve the body len once. For large message we may loop multiple times calling `read_in`.

Reply:
Got it, thanks for explaining!

I thought a bit about it, and correct me if I'm wrong, but this would also mean that when we read from a stream that can return `WouldBlock` (like in the case of `tokio-tungstenite`), we're still going to over-reserve because we'll potentially re-enter the `read_frame()` function multiple times essentially resetting the flag (`WoudBlock` may be returned by `read_in` in real-life but would not not happen in benchmarks due to the use of blocking sockets)?

If my assumption is correct, we could replace `if !reserved_full_msg_len` with something like `if self.in_buffer.capacity() < (len + header.size())` (so we'll eliminate the need for the flag while still ensuring that we don't over-reserve much more than needed).

Reply:
Yes `WouldBlock` during a large read will indeed currently cause over-reservation. But a capacity check won't help, that's just what reserve already does. If we've written 1MB of a 2MB message then hit WouldBlock the capacity could be 1MB and be enough. We'd want to avoid reserving 2MB in this case.

The code in this PR achieves this for error-less reads. To solve this for `WouldBlock` scenario we should store the flag in `self.header` or just do the reserve call when the header is populated.

Reply:
https://github.com/snapview/tungstenite-rs/pull/501